contracts/test/GracefulReturnTestHooks.sol

```diff
@@ -0,0 +1,221 @@
+        return IHooks.afterInitialize.selector;
+    }
+
+    function beforeModifyPosition(address caller, PoolKey calldata, IPoolManager.ModifyPositionParams calldata)
+        external
+        override
+        returns (bytes32)
+    {
+        CallerQueue memory previous = queue;
+
+        queue = CallerQueue({
+            caller: caller,
+            nonce: previous.nonce + 1,
+            stage: uint8(Stage.Modify),
+            head: keccak256(abi.encode(previous))
+        });
+
+        lastCaller = caller;
+
+        return Hooks.NO_OP;
+    }
+
+    function afterModifyPosition(address, PoolKey calldata, IPoolManager.ModifyPositionParams calldata, BalanceDelta)
+        external
+        pure
+        override
+        returns (bytes4)
+    {
+        return IHooks.afterModifyPosition.selector;
+    }
+
+    function beforeSwap(address caller, PoolKey calldata, IPoolManager.SwapParams calldata)
+        external
+        override
+        returns (bytes32)
+    {
+        CallerQueue memory previous = queue;
+
+        queue = CallerQueue({
+            caller: caller,
+            nonce: previous.nonce + 1,
+            stage: uint8(Stage.Swap),
+            head: keccak256(abi.encode(previous))
+        });
+
+        lastCaller = caller;
+
+        return Hooks.NO_OP;
+    }
+
+    function afterSwap(address, PoolKey calldata, IPoolManager.SwapParams calldata, BalanceDelta)
+        external
+        pure
+        override
+        returns (bytes4)
+    {
+        return IHooks.afterSwap.selector;
+    }
+
+    function beforeDonate(address caller, PoolKey calldata, uint256, uint256) external override returns (bytes32) {
+        CallerQueue memory previous = queue;
+
+        queue = CallerQueue({
+            caller: caller,
+            nonce: previous.nonce + 1,
+            stage: uint8(Stage.Donate),
+            head: keccak256(abi.encode(previous))
+        });
+
+        lastCaller = caller;
+
+        return Hooks.NO_OP;
+    }
+
+    function afterDonate(address, PoolKey calldata, uint256, uint256) external pure override returns (bytes4) {
+        return IHooks.afterDonate.selector;
+    }
+
+    function getHookSwapFee(PoolKey calldata key) external view override returns (uint8) {
+        return swapFees[key.toId()];
+    }
+
+    function getHookWithdrawFee(PoolKey calldata key) external view override returns (uint8) {
+        return withdrawFees[key.toId()];
+    }
+
+    function setReturnValue(bytes4 key, bytes4 value) external {
+        returnValues[key] = value;
+    }
+
+    function setSwapFee(PoolKey calldata key, uint8 value) external {
+        swapFees[key.toId()] = value;
+    }
+
+    function setWithdrawFee(PoolKey calldata key, uint8 value) external {
+        withdrawFees[key.toId()] = value;
+    }
+
+    function simulateHead(address caller, uint256 nonce, uint256 stage, bytes32 head) external pure returns (bytes32) {
```

Comment:
i must say..... i am very confused

Reply:
This is all from #291, I can remove it after I merge my fee changes, the test is not super necessary

Reply:
The original idea test concept that prompted my PR was using a sequencer inside the V4 hook, so some of that logic carried forward into the test.  My thinking was that sequencing the storage writes and using that to verify that *all* storage writes were successfully saved in the right order would help future-proof the test for transient storage... but i doubt it'll actually be relevant.